### PR TITLE
JAMES-2249 add cucumber scenario for JMAP sharing with subfolder

### PR DESCRIPTION
**Pull request by strokyl**
This commit add a integration test in JMAP to check that sharee can read shared
folder without having access of subfolder that are not shared of this folder.

### CODE REVIEW

**Comment by chibenwa** [comment]
```
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR] Failures: 
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR]   
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] Expected size:<1> but was:<0> in:
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] <[]>
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR] Scenario: Alice can list message from shared and shared/secret mailbox.Then the message list has size 1(Scenario: Alice can list message from shared and shared/secret mailbox)
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR]   Run 1: 
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] Expected size:<1> but was:<0> in:
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] <[]>
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [INFO]   Run 2: PASS
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [INFO] 
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR]   
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] Expected size:<1> but was:<0> in:
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] <[]>
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] [ERROR]   
[1c3d28b290d3e8aa2308bc215270335e9e3ac416] Expected size:<1> but was:<0> in:
```

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessageListMethodStepdefs.java [hunk 1]
```diff
@@ -72,6 +74,42 @@ public void getMessageList(String username, List<String> mailboxes, String flag)
             flag));
     }
 
+    @When("^\"([^\"]*)\" asks for message list in (?:mailboxes|mailbox) \"([^\"]*)\"$")
+    public void getMessageList(String username, List<String> mailboxes) throws Exception {
+        String mailboxIds = Joiner.on("\",\"")
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessageListMethodStepdefs.java]: The way to generate this _mailboxIds_ string seems duplicated everywhere.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    And the mailbox "shared" has 1 messages
+
+  Scenario: Alice can see shared mailbox and shared/secret mailbox
+    When "alice@domain.tld" lists mailboxes
+    And the mailboxes should contain "shared" in "Personal" namespace
+    And the mailboxes should contain "secret" in "Personal" namespace, with parent mailbox "shared" of user "alice@domain.tld"
+    And the mailbox "shared" has 1 messages
+    And the mailbox "secret" has 1 messages
+
+  Scenario: Alice can get message from shared and shared/secret mailbox
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature]: I don't think it's a good pattern to have multiple When/Then in a single scenario.

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+Feature: Share parent mailbox without sharing submailbox
+  As a James user
+  I want to be able to retrieve my mailboxes
+
+  Background:
+    Given a domain named "domain.tld"
+    And a user "alice@domain.tld"
+    And a user "bob@domain.tld"
+    And "alice@domain.tld" has a mailbox "shared"
+    And "alice@domain.tld" has a mailbox "caca"
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature]: hope you can find a better name

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    And "alice@domain.tld" has a mailbox "shared.secret"
+    And "alice@domain.tld" shares her mailbox "shared" with "bob@domain.tld" with "aeilrwt" rights
+    And "alice@domain.tld" has a message "m1" in "shared" mailbox
+    And "alice@domain.tld" has a message "m2" in "shared.secret" mailbox
+    And "alice@domain.tld" has a message "m3" in "caca" mailbox
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature]: here also

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithAndWithoutChildren.feature [hunk 5]
```diff
@@ -25,51 +26,80 @@ Feature: Share parent mailbox without sharing submailbox
...
+  Scenario: Bob can not get message from mailbox not shared with him
     When "bob@domain.tld" ask for messages "m2"
     And the list should contain 0 message
 
-  Scenario: Alice can list message from shared and shared/secret mailbox
+  Scenario: Alice can list message from all her shared mailbox
     When "alice@domain.tld" asks for message list in mailbox "shared"
     Then the message list has size 1
+    When "alice@domain.tld" asks for message list in mailbox "shared.notsecret"
+    Then the message list has size 1
```
Comment (rouazana) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithAndWithoutChildren.feature]: there are still scenarios with multiple when/then

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature (hunk 2 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithoutChildren.feature]: Not here anymore anyway

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessageListMethodStepdefs.java (hunk 1 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/java/org/apache/james/jmap/methods/integration/cucumber/GetMessageListMethodStepdefs.java]: Done, and yes it was messy and everywhere

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SharingParentMailboxWithAndWithoutChildren.feature [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    And the mailbox "shared" has 1 messages
+    And the mailbox "notsecret" has 1 messages
+
+  Scenario: Alice can see all her not shared mailbox
+    When "alice@domain.tld" lists mailboxes
+    And the mailboxes should contain "secret" in "Personal" namespace, with parent mailbox "shared" of user "alice@domain.tld"
+    And the mailbox "secret" has 1 messages
+
+  Scenario: Alice can get message from her shared mailbox
+    When "alice@domain.tld" ask for messages "m1"
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SharingParentMailboxWithAndWithoutChildren.feature]: Double When ;)

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SetFlagAndSharing.feature [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  Background:
+    Given a domain named "domain.tld"
+    And some users "alice@domain.tld", "bob@domain.tld"
+    And "alice@domain.tld" has a mailbox "shared"
+    And "alice@domain.tld" shares her mailbox "shared" with "bob@domain.tld" with "aeilrwt" rights
+    And "alice@domain.tld" has a message "m1" in "shared" mailbox
+
+  Scenario: Set flags by sharer should update unseen count when read by sharer
+    When "alice@domain.tld" sets flags "$Seen" on message "m1"
+    When "alice@domain.tld" lists mailboxes
```
Comment (aduprat) [inline server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SetFlagAndSharing.feature]: Replace the second When by And

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SharingParentMailboxWithAndWithoutChildren.feature (hunk 6 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SharingParentMailboxWithAndWithoutChildren.feature]: done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SetFlagAndSharing.feature (hunk 7 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/sharing/SetFlagAndSharing.feature]: done

File: server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithAndWithoutChildren.feature (hunk 5 above)
Comment (strokyl) [reply server/protocols/jmap-integration-testing/jmap-integration-testing-common/src/test/resources/cucumber/SharingParentMailboxWithAndWithoutChildren.feature]: done

**Pull request merged**